llvm/lib/Transforms/Utils/SimplifyCFG.cpp

```diff
@@ -1853,6 +1853,113 @@ static void hoistConditionalLoadsStores(
   }
 }
 
+static std::optional<bool>
+visitConditions(Value *V, const Value *baseCond, const BasicBlock *contextBB,
+                SmallVectorImpl<Instruction *> &impliedConditions,
+                const DataLayout &DL) {
+  if (!isa<Instruction>(V))
+    return std::nullopt;
+
+  Instruction *I = cast<Instruction>(V);
+  // we only care about conditions in the same basic block
+  if (contextBB != I->getParent())
+    return std::nullopt;
+
+  std::optional<bool> Imp = isImpliedCondition(V, baseCond, DL);
+  // TODO: Handle negated condition case.
+  if (Imp != true)
+    return std::nullopt;
+
+  std::optional<bool> LHS = visitConditions(I->getOperand(0), baseCond,
```

Comment:
I think this code is assuming that the instruction is an `and` without actually checking that?

Please add a test with e.g. `or`.

Reply:
I had a wrong understanding of what implication relation means, I have fixed it now and will work on adding more testcases.